src/DIRAC/Resources/Computing/CloudComputingElement.py

```diff
@@ -0,0 +1,549 @@
+            imageName = rawID[5:]
+        drv = self._getDriver()
+        for image in drv.list_images():
+            if not imageName and image.id == rawID:
+                return image
+            elif imageName and image.name == imageName:
+                return image
+        raise KeyError("No matching image found for %s" % rawID)
+
+    def _getFlavor(self):
+        """Extracts flavor from configuration system.
+
+        :return: flavor object
+        """
+        rawID = self.ceParameters.get("Instance_Flavor", None)
+        if not rawID:
+            raise KeyError("Flavor not set in Configuration")
+        flavorName = None
+        if rawID.startswith("name:"):
+            flavorName = rawID[5:]
+        drv = self._getDriver()
+        for flavor in drv.list_sizes():
+            if not flavorName and flavor.id == rawID:
+                return flavor
+            elif flavorName and flavor.name == flavorName:
+                return flavor
+        raise KeyError("No matching flavor found for %s" % rawID)
+
+    def _getSSHKeyID(self):
+        """Extract ssh key id from configuration system.
+
+        :return: ssh id string or None if not set
+        """
+        return self.ceParameters.get("Instance_SSHKey", None)
+
+    def _getMetadata(self, executableFile):
+        """Builds metadata from configuration system, cloudinit template
+         and dirac pilot job wrapper
+
+        :param str executableFile: path to pilot wrapper script to include in metadata.
+
+        :return: instance specific metadata in mime string format
+        """
+
+        default_template = os.path.join(os.path.dirname(__file__), "cloudinit.template")
+        template_file = self.ceParameters.get("Context_Template", default_template)
+        template_params = {}
+        template = ""
+        # Proxy needs extra indents to keep yaml formatting happy
+        template_params["proxy"] = self.proxy.replace(b"\n", b"\n     ").decode()
+        with open(executableFile) as efile:
+            raw_str = efile.read().strip()
+            raw_str = raw_str.replace("\n", "\n     ")
+            template_params["executable"] = raw_str
+        with open(template_file) as template_fd:
+            template = template_fd.read()
+            template = template % template_params
+
+        userData = MIMEMultipart()
+        mimeText = MIMEText(template, "cloud-config", sys.getdefaultencoding())
+        mimeText.add_header("Content-Disposition", 'attachment; filename="pilotconfig"')
+        userData.attach(mimeText)
+        return str(userData)
+
+    def _renewCloudProxy(self):
+        """Takes short lived proxy from the site director and
+        promotes it to a long lived proxy keeping the DIRAC group.
+
+        :returns: True on success, false otherwise.
+        :rtype: bool
+        """
+        if not self._cloudDN or not self._cloudGroup:
+            self.log.error("Could not renew cloun proxy, DN and/or Group not set.")
```

Comment:
```suggestion
            self.log.error("Could not renew cloud proxy, DN and/or Group not set.")
```

Reply:
Compared to VMDIRAC there is no more possibility to define multiple VMTYPE's, i.e. different images/sizes for VMs. I think this feature is useful and can be relatively easily implemented. To keep symmetry with other generic Computing Elements these VMTYPE's can be configured as different Queues.

Reply:
Yes, using queues for different instance types was exactly our intention.